accepted/future-releases/nnbd/number-operation-typing.md

```diff
@@ -62,47 +62,76 @@ T add<T extends num>(T a, T b) => a + b;
 
 ### Improved context type
 
 We extend type inference to take the special number rules into account.
 
-> For  `e1 + e2`, `e1 - e2`, `e1 * e2`, `e1 % e2` or `e1.remainder(e2)` where the static type of  `e1` is a non-`Never` subtype of `int`, if the context type of the entire expression is `int`, then the context type of `e2` is `int`. If the static type of `e1` is a non-`Never` subtype of `num` that is *not* a subtype of `double` and the context type of the entire expression is `double`, then the context type of `e2` is `double`.
+> For an expression `e` of the form  `e1 + e2`, `e1 - e2`, `e1 * e2`, `e1 % e2` or `e1.remainder(e2)`, if the static type of  `e1` is a non-`Never` subtype of `int`, if the context type of `e` is a non-`Nevers` subtype of `int`, then the context type of `e2` is `int`. If the static type of `e1` is a non-`Never` subtype of `num` that is *not* a subtype of `double` and the context type of`e` is `double`, then the context type of `e2` is `double`.
 
-> If the context type of `e1.clamp(e2, e3)` is a non-`Never` subtype of `num`, *T*, and the static type of `e1` is a non-`Never` subtype of *T*, then the context types of `e2` and `e3` are both *T*.<br>Otherwise, if the context type of the invocation is `int` and the static type of `e1` is a non-`Never` subtype of `int`, then the context types of `e2` and `e3` are both `int`, and if the context type of the invocation is `double` and the static type of `e1` is a non-`Never` subtype of `double`, then the context types of `e2` and `e3` are both `double`. Otherwise the context types of `e2` and `e3` are `num`.
+> If the context type of `e1.clamp(e2, e3)` is a non-`Never` subtype of `num`, *T*, then the context types of `e2` and `e3` are also *T*. Otherwise the context types of `e2` and `e3` are `num`.
 
 (This does emphasize the inherent non-symmetry of Dart operators: The first operand is a receiver which is always evaluated with no type context, and is then used to resolve the operator method against, and the second operand is an argument to that method. We need to fully resolve the first operand and the operator before we can even begin with the second operand.)
 
+Tabulated, the context type of the second operand, based on the first operand and the context type of the entire operation, is:
+
+| op1 \\ ctx  | int             | int+            | double | double+         | num  | num+            |
+| :---------: | --------------- | --------------- | ------ | --------------- | ---- | --------------- |
+|   **int**   | int             | int<sup>*</sup> | double | num<sup>*</sup> | num  | num<sup>*</sup> |
+|  **int+**   | int             | int             | double | num<sup>*</sup> | num  | num<sup>*</sup> |
```

Comment:
Column 3 (context `int+`) should be `int<sup>*</sup>` as well, I think: `e1 + e2` will have type `int`, not anything denoted by `int+`.

Reply:
if `e1` and `e2` have the same type, then the result has that type.

So, if both `e1` and the context type are the same a proper subtype of `int`, then so must `e2` be. So, it should be `int+`.

Reply:
OK, it is actually sound: We could for instance have a receiver whose static type is a type variable `X` and dynamic type _S_ is "JavaScript number" which is a proper subtype of `int`, but the result would also be of type _S_. So even if the actual value of `X` could somehow be _S_ it is still sound. Similarly, the treatment of `Never` is also sound.